build/ci.go

```diff
@@ -345,11 +349,77 @@ func downloadSpecTestFixtures(csdb *build.ChecksumDB, cachedir string) string {
 	return filepath.Join(cachedir, base)
 }
 
+// hashDir iterates all files under the top-level project directory
+// computing the hash of each file (excluding files within the tests
+// subrepo)
+func hashDir() (map[string]common.Hash, error) {
+	res := make(map[string]common.Hash)
+	err := filepath.WalkDir(".", func(path string, d os.DirEntry, err error) error {
+		if strings.HasPrefix(path, filepath.FromSlash("tests/testdata")) {
+			return nil
+		}
+		if d.IsDir() {
+			return nil
+		}
+		// open the file and hash it
+		f, err := os.OpenFile(path, os.O_RDONLY, 0666)
+		if err != nil {
+			return err
+		}
+		contents, err := io.ReadAll(f)
+		if err != nil {
+			return err
+		}
+		res[path] = common.Hash(sha256.Sum256(contents))
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+var cachedir = flag.String("cachedir", "./build/cache", "directory for caching golangci-lint/protoc binaries.")
+
+// doGenerate ensures that re-generating generated files does not cause
+// any mutations in the source file tree:  i.e. all generated files were
+// updated and committed.  Any stale generated files are updated.
+func doGenerate() {
+	tc := new(build.GoToolchain)
+
+	protocPath, err := downloadProtoc(*cachedir)
+	if err != nil {
+		log.Fatalf("error downloading protoc", "err", err)
+	}
+
+	preHashes, err := hashDir()
+	if err != nil {
+		log.Fatal("failed to compute map of source hashes", "err", err)
+	}
+
+	c := tc.Go("generate", "./...")
+	newPathEnv := fmt.Sprintf("%s:%s/bin", os.Getenv("PATH"), protocPath)
```

Comment:
Please use `filepath.PathListSeparator` and `filepath.Join` here to make this work on Windows too. Probably would make sense to create the value with `+` instead of `fmt.Sprintf`.